### ISSUE TITLE
Fine tune L2G after feature matrix updates

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/issues#3910]
As a **developer** I want **to rerun the model parametrisation pipeline** because **the latest changes in the feature matrix require finding the best set of parameters that adapt to the new data patterns.**

## Background

This step occurs outside of Gentropy due to the particularities of the infrastructure. We perform L2G training on a parameter grid (almost 3000 combinations) and we run this process in parallel on a large machine.
The notebook I have run is here https://gist.github.com/ireneisdoomed/7b425900f6df57b24e4f30bfe914c744

## Tasks

- [x] Create big cluster to parallelise training (`n4-highcpu-64`) and with access to a Jupyter server
- [x] Connect to the machine and run the code in the Gist above
- [x] Collect all results in a spreadsheet - Results are here https://docs.google.com/spreadsheets/d/1WrV6upVyn3IiihlVQkSOg8s1lCX___IJOIe0Cb2UMoI/edit?usp=sharing

## Review

As usual, there are several sets of params that give good results. I've picked the one at the top:
```
{'ccp_alpha': 0.0, 'learning_rate': 0.1, 'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 10, 'subsample': 1.0, 'tol': 0.001}
```

I've validated the results with cross validation, to make sure they are robust to several splits. Variability is low between them:
```
Performing cross-validation...
Fold 1: Average Precision = 0.920
Fold 2: Average Precision = 0.902
Fold 3: Average Precision = 0.895
Fold 4: Average Precision = 0.942
Fold 5: Average Precision = 0.919

Cross-validation Average Precision: 0.916 (+/- 0.033)

Training final model on full training set...

Evaluating on held-out test set...
Test Set Average Precision: 0.925
```

**Something important to take into account is that the model complexity has increased**, presumably after downweighting the importance of the distance features, making the task more complicated to resolve.
All top results have a max depth tree of 7 and 10. This is the topology of one of the trees:

![Image](https://github.com/user-attachments/assets/a95b3a1d-fd94-41f9-b4d8-1c8c433b535e)

We should expect an impact in the performance when calculating L2G probabilities. If it is a problem, I suggest using these params `{'ccp_alpha': 0.0, 'learning_rate': 0.1, 'max_depth': 5, 'min_samples_leaf': 10, 'min_samples_split': 10, 'subsample': 1.0, 'tol': 0.001}`. With a precision of 0.916, it is the model that performs the best in comparison with its topology.

**Comment by d0choa** [comment]
Let's see if this works. Complexity caused trouble in the calculation of the shapely values in the past. 🤞

**Comment by project-defiant** [comment]
This seems likely because of the `max_depth`

**Comment by ireneisdoomed** [comment]
@addramir The splitting follows the same principle as before. There is one major difference, though: since each effector gene is now associated with a list of traits, rather than a single EFO, I consider all traits to contribute to the gene-disease association as we discussed.

This can have an effect when splitting by groups of gene/EFO so that credible sets from these 2 effector genes:
- Gene A / Anaemia and Kidney Disease (from [GCST90475792](https://www.ebi.ac.uk/gwas/studies/GCST90475792))
- Gene A / Anaemia 
end up in the train and test set. Data leakage between splits would overinflate the evaluation metrics.

To avoid this, an option would be to explode the list of traits so that we form the following pairs:
- Gene A / Anaemia
- Gene A / Kidney Disease
I think this is the approach we used to take. The drawback is that we'd decouple traits that are studied together in a study.


## Disease annotation analysis
```
Gold Standard Size: 103,031
Credible sets with multiple disease IDs: 44,975
---
Total unique genes: 4,756
Genes with multiple disease combinations: 2,658
Genes with subset relationships (potential leakage): 2,012
Percentage of genes with potential leakage: 42.30%
```

I call a subset relationship those cases where gene 1 is linked to [disease1, disease2] by one study, and to [disease1] by another study.
Based on this, the risk of data leakage is quite high. **I suggest splitting based on gene IDs only**, it is more conservative.

**Comment by addramir** [comment]
I think that exploding of the diseaseIds list is not a correct way to do it.
To my opinion, if study has multiple EFOs - this combination should be considered as unique EFO because it usually reflects unique design or special trait. 
So in case of the list like this:
gene1-EFO1
gene1-EFO2
gene1-EFO1+EFO2
All of them should be treated as unique gene-EFO combination.  

I don't think it is reasonable split based on geneID only because it will make the effective size of the training set much smaller.

**Comment by d0choa** [comment]
@ireneisdoomed shall we close?

**Comment by ireneisdoomed** [comment]
https://github.com/opentargets/gentropy/pull/1076 implemented a custom splitting methodology to avoid data leakage
